### ISSUE TITLE
Support different request type in one ABAC model

### ISSUE THREAD

**Issue by kbsiva** [casbin/Casbin.NET#108]
Hi All, I am using below ABAC config and I have defined the below two rules in the database:

```ini
[request_definition]
r = sub, obj, act

[policy_definition]
p = sub_rule, obj, act

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = eval(p.sub_rule) && r.obj == p.obj && r.act == p.act
```

Rules In Databas:

```
Id       PType      V0                                                                 V1        V2    V3      V4      V5
2070 p             r.sub.Age >=18 && r.sub.Age <=60             data1 read NULL NULL NULL
2071 p              r.sub.prodid==2018 && r.sub.version==58 data1 read NULL NULL NULL
```

```c#
var efCoreAdapter = new CasbinDbAdapter<int>(context);
var enforcer = new Enforcer("examples/abac_model.conf", efCoreAdapter);
enforcer.LoadPolicy();
Person p1 = new Person { Age = 19 };
enforcer.Enforce(p1, "data1", "read"); // This returns true as it maches the policyid 2070

// Similarly I defined another rule 2071 with attributes as prodid and versionid
        productdetails pdetails = new productdetails { prodid = 2018, version = 58 };
         enforcer.Enforce(pdetails, "data1", "read"); // This retuns below exception
System.ArgumentException: 'Object of type 'productdetails' cannot be converted to type 'Person'.'
```

Why it is taking Person as the default sub attribute and not taking the rule that is mentioned for productdetails? How to switch of casbin matches the subrule based on the roles policies defined in the table.?

**Comment by Sagilio** [comment]
Because the same lambda will be cached for each different expression and the request type will be set depending on the first policy. 
In fact, Enforce will match the policies in order. Therefore, even without caching, A similar exception will be thrown for different request types.
```
DynamicExpresso.Exceptions.ParseException: 'No property or field 'Age' exists in type 'ProductDetails' (at index 6).'
```
You can write two model and policy files for Person and ProductDetails types to solve this question temporarily and we will try to provide a better API design on PR #94 to resolve this issue.

**Comment by hsluoyz** [comment]
@Sagilio can you see how Golang Casbin implemented this? I didn't remember that we have such a limitation. Anyway, I think this can be finally resolved by reflection at run-time.

**Comment by Sagilio** [comment]
@hsluoyz The go casbin do not cache the expression (EvaluableExpression). So we can also fix this issue by disabling the expression cache when using the ABAC model.
The previous [comment](https://github.com/casbin/Casbin.NET/issues/108#issuecomment-766319910) means we can cache different request types of expressions to fix this issue, It has minimal impact on performance.

**Comment by hsluoyz** [comment]
@Sagilio 

> The previous comment means we can cache different request types of expressions to fix this issue, It has minimal impact on performance.

This way sounds good to me.

**Comment by sagilio** [comment]
closed by #230 , now we will use a different cache when request type is different
